feature/bioformats-reader

tlambert03 opened the pull request:
## Description
This PR implements a (java-backed) bioformats reader via jpype, with the jar coming from [`bioformats_jar`](https://github.com/tlambert03/bioformats_jar).

This requires `bioformats_jar` (which will install jpype), and also requires that the user has java in their environment.  (The easiest way to do that in a python env is to use `conda install -c conda-forge openjdk`).

From discussions, I think we all see this as a fallback reader, with the ultimate goal still being to provide pure python readers.  But in the meantime, this brings all of the bioformats supported formats into the same `Reader` API here, and also adds a dask-backed bioformats reader, which I believe might be new?

The classes added here are:

- `aicsimageio.readers.bioformats_reader.BioFile` - the lowest level reader object, with an API somewhat similar to TiffFile.  can be used as a context manager (`with BioFile('path') as f: ...`) and offers these public attributes:
   - `set_series`
   - `core_meta` - (named tuple with core metadata)
   - `open/close`
   - `to_numpy` - greedy reader (always `TCZRX[c]`, where `[c]` is for RGB)  
   - `to_dask` - delayed reader (always `TCZRX[c]`) 
   - `ome_metadata` - returns parsed metadata as `ome_types.OME`
   - `ome_xml` - returns OME XML string
- `aicsimageio.readers.bioformats_reader.BioformatsReader` - wraps a `BioFile` and implements the Reader protocol. 

This also adds a convenience function `aicsimageio.utils.ome_utils.bioformats_ome` that can be used to retrieve `ome_types.OME` from any bioformats supported file.  (This was the original motivation for https://github.com/tlambert03/ome-types/pull/102 ... and allows me to close that)

## Pull request recommendations:
- [x] Name your pull request _your-development-type/short-description_. Ex: _feature/read-tiff-files_
- [x] Link to any relevant issue in the PR description. Ex: _Resolves [gh-12], adds tiff file format support_
   - relevant: https://github.com/AllenCellModeling/aicsimageio/issues/147
- [x] Provide relevant tests for your feature or bug fix.
- [x] Provide or update documentation for any feature added by your pull request.

JacksonMaxfield commented:
@tlambert03 pretty sweet to even see an early implementation so quickly! Want to me review as commits come in or just ping me when to review? :joy:

tlambert03 commented:
No need just yet.  There if you're curious, but can ping when closer

tlambert03 commented:
coming along... I created a new package: https://github.com/tlambert03/bioformats_jar
this contains just the jar and the bare minimum jpype boilerplate.  (in theory, it's something that everyone could use: bfio, pims).  I found an amazing package ([stubgenj](https://gitlab.cern.ch/scripting-tools/stubgenj)) that generates type stubs from jpype.  So if you do `from bioformats_jar import loci`, you'll get stubs and autocompletion for the loci API.  (there's some kinks to iron out, but it's mostly there and it's awesome).

remaining issues:
- stray file pointers.  the `check_local_file_not_open` test fails a lot here... even if I'm not saving a python reference to a wrapped ImageReader.  So I need to track down what's going on with file descriptors on the java side.
- multi-scene files are still not implemented
- more code cleanup

tlambert03 commented:
(@JacksonMaxfield, can you approve the CI workflow again?  I'm curious to see if the tests that pass locally are passing on CI)

tlambert03 commented:
ping @JacksonMaxfield  ... another test please?

this is getting closer.  lots of tests passing locally here.  You could start taking a look if you want

tlambert03 commented:
cool!  got a couple full test suite passes (macos-11, 3.9 & 3.8).  many of the other fails are because it failed to find java or because of some linux threading issues

(side discussion to have:  do you want to add the `setup-java` to the main tests as I've done here?  or create a separate job for bioformats tests?  I don't think it adds to much time/overhead beyond the tests themselves)

JacksonMaxfield reviewed (commented):
This is all _very_ impressive. Especially the dask array wrapper to open the file during compute calls. We may want to export that to it's own file at some point because I can foresee other readers wanting to use it.

Generally I don't have any major comments. If you point me in the direction of some test files for more formats would be happy to add them to the test dataset so you can add tests specific to ND2, and more. At the very least I would love to see these tests run against CZI and LIF files too. Because of that, I would say you can remove some of the tests. Maybe just one test per file type? One tiff, one OME-TIFF, one CZI, one LIF, one ND2, etc.

JacksonMaxfield commented:
Also pinging @toloudis. Time to start reviewing 🙂

tlambert03 commented on aicsimageio/readers/bioformats_reader.py:
```diff
@@ -0,0 +1,457 @@
+
+        return xr.DataArray(
+            image_data,
+            dims=dimensions.DEFAULT_DIMENSION_ORDER_LIST_WITH_SAMPLES
+            if rgb
+            else dimensions.DEFAULT_DIMENSION_ORDER_LIST,
```
yeah, I also need to work on pyramid data... (currently skipped in the tests)

JacksonMaxfield replied:
We have a long term plan for pyramid data with a `set_level` API but haven't even start it (https://github.com/AllenCellModeling/aicsimageio/issues/175#issuecomment-765648637). For now feel free to just always select the lowest level.

JacksonMaxfield commented:
> cool! got a couple full test suite passes (macos-11, 3.9 & 3.8). many of the other fails are because it failed to find java or because of some linux threading issues
> 
> (side discussion to have: do you want to add the `setup-java` to the main tests as I've done here? or create a separate job for bioformats tests? I don't think it adds to much time/overhead beyond the tests themselves)

Feel free to add the setup java call to the build main workflow. the overhead is tiny iirc.

tlambert03 commented on aicsimageio/readers/bioformats_reader.py:
```diff
@@ -0,0 +1,457 @@
+        self.open()
+        self._lock = Lock()
+
+        if "ND2" in str(self._r.getReader()):
+            # https://github.com/openmicroscopy/bioformats/issues/2955
+            mo = loci.formats.in_.DynamicMetadataOptions()
+            mo.set("nativend2.chunkmap", "False")
+            self._r.setMetadataOptions(mo)
```
yeah :/ i agree... there may be lots of these.  The scope of bioformats is so massive... i'm sure if people actually use it we'll dig up lots of bugs, but maybe better to have and have frustrated than never to have had at all? :joy:

Side note: we can actually turn on/off support for individual readers as a preference if we want (i.e. select a subset from the [default list here](https://github.com/ome/bioformats/blob/develop/components/formats-api/src/loci/formats/readers.txt)).

JacksonMaxfield replied:
Totally agree. Better to have than have not and totally fine with adding in patches like this as we go (although hopefully we can get to a point where the person reporting the patch can just make the fix).

JacksonMaxfield replied:
Did not know about the ability to enable/disable readers. That's pretty neat.

joshmoore replied:
btw, there aren't THAT many options yet. ;)

https://docs.openmicroscopy.org/bio-formats/6.7.0/formats/options.html?highlight=options

tlambert03 replied:
ps @joshmoore by "we'll dig up lots of bugs" I meant in my implementation of wrapping bioformats more than in bioformats itself!  just an apologetic preamble 😂

tlambert03 commented on aicsimageio/readers/bioformats_reader.py:
```diff
@@ -0,0 +1,457 @@
+
+    def to_numpy(self, series: Optional[int] = None) -> np.ndarray:
+        """Create numpy array for the current series."""
+        return np.asarray(self.to_dask(series))
```
yep, pretty much just calls compute [(code here)](https://github.com/dask/dask/blob/0530bdbf641f6cd33618ef648ad737a4a3d48d4e/dask/array/core.py#L1490-L1496) ... but I agree, it's a nicer idiomatic way to say "to numpy please"

```py
    def __array__(self, dtype=None, **kwargs):
        x = self.compute()
        if dtype and x.dtype != dtype:
            x = x.astype(dtype)
        if not isinstance(x, np.ndarray):
            x = np.array(x)
        return x
```

tlambert03 commented on aicsimageio/readers/bioformats_reader.py:
```diff
@@ -0,0 +1,457 @@
+        # if someone indexes a 5D dask array as `arr[0,1,2]`, then the `info`
+        # dict will contain the key: 'chunk-location': (0, 1, 2, 0, 0)
+        # reader.getIndex() expects (Z, C, T) ...
+        # We ASSUME that the final dask array is in the order TCZYX, so chunk-location
```
for now, that's an internal assumption (and it just makes things easier).  as far as our interaction with bioformats is concerned, it all really comes down to `openBytes` call [here](https://github.com/AllenCellModeling/aicsimageio/pull/306/files#diff-2d684fcc182480aea8c37c6dc033094a287076307dd0796601d42d1a7e281259R319) where you just give it an ZCT index and it gets it for you.  As we discussed, I'm honestly not quite sure what the relevance of `getDimensionOrder` is.  that is: it doesn't impose limitations on how you retrieve the planes.  So for now I've just been saying "stick with TCZYX", ignore native dimensionOrder (until we learn why we might not want to), and let the end-user reshape our xarray if they want to

JacksonMaxfield replied:
I think the problem with that is that our most common interaction is `get_image_data` / `get_image_dask_data` which takes in a dimension string instead of integer / slice indices.

if the backing data has the data mapped to dimensions incorrectly then I would assume the `get_image_data` would return incorrectly mapped data too.

JacksonMaxfield replied:
Maybe I am misunderstanding though. Are you storing the data in our native order or in the files native order? If you are storing the data in our native order then the above concern is invalid.

tlambert03 replied:
let me look into it a bit.  my intention was to provide the data always in the native order... as a 5d  TCZYX array if not RGB or a 6D `TCZYXc` if rgb.  I basically ignore the file's native order.

toloudis replied:
I think it's totally reasonable that a Bioformats reader would give back TCZYX because it is itself a layer on top of proprietary file format readers.  Just as long as the metadata is consistent with that...

JacksonMaxfield replied:
> my intention was to provide the data always in the native order... as a 5d TCZYX array if not RGB or a 6D TCZYXc if rgb. I basically ignore the file's native order.

that is totally fine with me 👍

tlambert03 commented on aicsimageio/readers/bioformats_reader.py:
```diff
@@ -0,0 +1,457 @@
+        # reader.getIndex() expects (Z, C, T) ...
+        # We ASSUME that the final dask array is in the order TCZYX, so chunk-location
+        # will be coming in as (T, C, Z, Y, X). `[2::-1]`` converts that to (Z, C, T)
+        # TODO: we could either use the native getDimensionOrder? (or not) or let
+        # the user
```
see above...

tlambert03 commented on aicsimageio/utils/cached_property.py:
```diff
@@ -0,0 +1,62 @@
+
+    _NOT_FOUND = object()
+
+    # backport from 3.8
```
yep, honestly this is optional too (I could remove it if you want and we can just wait for 3.8).  It's seems like a pretty handy util for this library, since there are a lot of lazy props that only needed to be calculated once.

toloudis replied:
I think we would rather have this earlier and still support 3.7 but agree this needs a louder comment saying why it exists

toloudis commented on aicsimageio/formats.py:
```diff
@@ -194,4 +195,7 @@
     "lfp": "aicsimageio.readers.default_reader.DefaultReader",
     "spe": "aicsimageio.readers.default_reader.DefaultReader",
+
+    "nd2": "aicsimageio.readers.bioformats_reader.BioformatsReader",
```
what does this look like when the full implementation is ready? It's making me imagine things like BioformatsReader.supported_extensions : List[str]

JacksonMaxfield replied:
I think we are still waiting for a decision / conclusion on: https://github.com/AllenCellModeling/aicsimageio/issues/307

joshmoore replied:
see possible endings (though note that's not always sufficient) which can also be introspected with Java: http://j.mp/bio-formats-list

JacksonMaxfield replied:
We should figure this out before this is merged imo. Does anyone on the review list have hard opinions on the resolution of #307... @tlambert03 do you have hard opinions on that one? I haven't tested this reader out at all but in your testing of it what were your gripes.

JacksonMaxfield replied:
And regardless of what our resolution strategy is, all of the formats for bioformats should be unpacked to this list before merge.

tlambert03 replied:
> do you have hard opinions on that one?

not really.  probably leaning slightly to the one-reader-per-extension as @toloudis suggested.  Since ultimately that list is consumed by the AICSImage object (right?) having the additional overhead of a fallback reader seems excessive – at least at first – since a user could still directly indicate the backend if they wanted to override, (correct?)

JacksonMaxfield replied:
Ahhhh true. I forgot that the `reader=` param exists on the `AICSImage` object so having one reader per format is just a totally fine default.

tlambert03 replied:
> I haven't tested this reader out at all but in your testing of it what were your gripes.

main gripe is probably the time to load (at first).  Then it speeds up a bit.  but for things that it attempts, it feels pretty robust.  that is: as I started playing with other files, they tended to "just work" without any special considerations.  which I guess is how it "should" be given that most of this is tested on the bioformats side?

toloudis replied:
When merging / accepting this PR, I don't think we would consider it "experimental" at all.  There may or may not be use cases that are non optimal or don't work but by including it we are basically saying we think the 80% case is working well.  The intent would be to support it moving forward.

toloudis replied:
And let me ALSO say a thanks to @tlambert03 here.  This is a very high impact update.  

The very early versions of this library (well prior to it arriving on github, and also pre-daskification) tried to use python-bioformats from CellProfiler and we ran into lots of issues internally with multiple java vms running at the same time, crashes etc.   We (AICS) felt that a more "pure" python solution (resulting in using tifffile and czifile inside the Readers in those earliest versions) was going to be more robust and that was the case for us.  

But now is the perfect time to add this support.  We do still prefer more "native" python format readers but having this type of universal fallback is just perfect -- so also big thanks to the OME for BioFormats even existing.  As I said in a prior discussion, we now have something where before there was nothing.

toloudis commented on aicsimageio/readers/bioformats_reader.py:
```diff
@@ -0,0 +1,457 @@
+        )
+
+    def _to_xarray(self, delayed: bool = True) -> xr.DataArray:
+        # TODO: put in a utils method somewhere?
```
yes

toloudis commented on aicsimageio/readers/bioformats_reader.py:
```diff
@@ -0,0 +1,457 @@
+    @property
+    def ome_metadata(self) -> "OME":
+        """Return OME object parsed by ome_types."""
+        import ome_types
```
is this stuff expensive to just import at the top of the file?

tlambert03 replied:
nope, just a habit i suppose.  particularly that one which is used elsewhere in this repo  (and therefore almost guaranteed to be imported already).  will move them

JacksonMaxfield commented:
@tlambert03 I got message on zulip but can't sign in on mobile for some reason.

Feel free to add pre-commit.

tlambert03 commented:
lol :) ok thanks

tlambert03 commented:
ok!  I think I figured out the main segfault issues with windows and linux, and might even have saved a buffer copy in the process.  Everything but the docs tests seem to be passing now.  I did some general cleanup and documentation, and adde full test coverage on that experimental dask array proxy (I think it's promising... but still possibly error prone?)

As a first implementation, I think this is nearly ready.  Have a look at your convenience, and let's figure out how we want to handle formats.py

tlambert03 commented:
> If you point me in the direction of some test files for more formats would be happy to add them to the test dataset so you can add tests specific to ND2, and more. At the very least I would love to see these tests run against CZI and LIF files too. Because of that, I would say you can remove some of the tests. Maybe just one test per file type? One tiff, one OME-TIFF, one CZI, one LIF, one ND2, etc.

still need to do this... there's tons of images at https://downloads.openmicroscopy.org/images/ ... i'll pick a selection and have you add them to the test dataset

tlambert03 commented:
ok I added tests against a number of file types:
- ome-tiff
- czi
- lif
- nd2 (nikon)
- dv & r3d (deltavision)
- oir (olympus)
- ims (imaris)
- klb (keller lab binary)
- dcm (dicom)

I'm grabbing that data from https://downloads.openmicroscopy.org/images/ ... have a look in `download_test_resources` and let me know if you'd prefer to move to s3

tlambert03 commented:
all checks are green!  I updated the original post to provide a little more detail and context for anyone finding this PR.  I'll wait for another round of review now before doing anything else here.

JacksonMaxfield commented on aicsimageio/utils/ome_utils.py:
```diff
@@ -0,0 +1,145 @@
+"""
```
nitpick: can we move this entire file to the metadata module: https://github.com/AllenCellModeling/aicsimageio/tree/main/aicsimageio/metadata

or the content of this file to `metadata/utils`? Was there a reason for _not_ doing that?

tlambert03 replied:
nope, will do

JacksonMaxfield commented on scripts/download_test_resources.py:
```diff
@@ -101,13 +107,40 @@ def download_test_resources(args: Args):
+        log.info(str(ROOT + suffix))
+        urlretrieve(str(ROOT + suffix), dest)
+
+    files = [
+        "ND2/aryeh/but3_cont200-1.nd2",
+        "ND2/jonas/header_test2.nd2",
+        "ND2/maxime/BF007.nd2",
+        "DV/siRNAi-HeLa/IN_02.r3d_D3D.dv",
+        "DV/siRNAi-HeLa/IN_02.r3d",
+        "Olympus-OIR/etienne/amy%20slice%20z%20stack_0001.oir",
+        "Imaris-IMS/davemason/Convallaria_3C_1T_confocal.ims",
+        "KLB/samples/img.klb",
+        "DICOM/samples/MR-MONO2-8-16x-heart.dcm",
```
I will add these all to the package today. Mind if I push a commit to update the TEST_RESOURCES_HASH after I do? / If I can't push to this PR branch then I will just comment with the new hash.

tlambert03 replied:
thanks! 
> Mind if I push a commit to update the TEST_RESOURCES_HASH after I do

of course not.  go for it.

JacksonMaxfield replied:
Don't think I have push access, new test resources hash: `9008c2bff8a87f7fecf6f1b111b6d8c1f8f7b49756939a84efc5e57ad878f017`

tlambert03 replied:
thanks, added and reverted changes to download script

JacksonMaxfield reviewed (commented):
Some minor changes requested on my end but I looked through all of this and really couldn't see anything major. Excited to see this come through so quickly!

Will wait for @toloudis and one other to review as well.

A general note -- you should add the new reader in some fashion to the README format support (likely similar to how I did the "base-imageio" one) but with a few more directions on setting Java likely.

JacksonMaxfield commented:
After seeing the checks kick off, I realized you should probably include the workflow changes you made on `test-and-lint` to `build-main`

sbesson commented:
Wow a lot going of work going here. Big :+1: to @tlambert03 for the wrapping . 

Two high-level questions:
- does the GPL licensing of some of the Bio-Formats reader cause a potential issue as this repo is BSD as far as I can see? Or is the downloading of the JAR effectively independent of the installation of the library so it is rather a runtime dependency ?
- do you have a timeline for this functionality to be released and a general feeling for how experimental/supported the feature will be? The question of Bio-Formats in Python is regularly asked e.g. on image.sc. I assume on the OME side we would be more than happy to point at `aicsimageio` with this change integrated as one of the solutions but given the exposure I would definitely not do that without your consent ;)

JacksonMaxfield commented:
> Wow a lot going of work going here. Big +1 to @tlambert03 for the wrapping .

Fully agree, @tlambert03 this is incredible and it should be said more often haha.

> Two high-level questions:
> 
>     * does the GPL licensing of some of the Bio-Formats reader cause a potential issue as this repo is BSD as far as I can see? Or is the downloading of the JAR effectively independent of the installation of the library so it is rather a runtime dependency ?

This is an interesting one. I _believe_ we can license each module separately? I.e. "if you are using the CZI, LIF, or BioFormats Readers your inheritting library or scripts is automatically licensed under GPLv3"

My reasoning here is that they are all pretty separate readers. Like yes they live in the same repo but they aren't imported by default and they aren't even installed by default.

Which basically leaves TIFF + OME-TIFF + (future) Zarr as BSD3 licensed. :slightly_frowning_face: 

I honestly don't know though. I need to talk to Allen Legal I feel like.

>     * do you have a timeline for this functionality to be released and a general feeling for how experimental/supported the feature will be? The question of Bio-Formats in Python is regularly asked e.g. on image.sc. I assume on the OME side we would be more than happy to point at `aicsimageio` with this change integrated as one of the solutions but given the exposure I would definitely not do that without your consent ;)

Another great question, I would say "as soon as possible", given #307 and the legal stuff is resolved. Would definitely release and annouce it as "experimental" first (and I would say Talley should be the one to announce it considering he did the work :slightly_smiling_face: )

I will say, I am excited at the opportunity for this + #300 to both merge in so that we can have an "in Python" two liner for converting to OME

```python
from aicsimageio import AICSImage

img = AICSImage("some-file.nd2", reader=BioformatsReader)
img.save(...)
```

tlambert03 commented:
> Wow a lot going of work going here. Big 👍 to @tlambert03 for the wrapping .

👋 thanks for checking in @sbesson!

>  does the GPL licensing of some of the Bio-Formats reader cause a potential issue as this repo is BSD as far as I can see? Or is the downloading of the JAR effectively independent of the installation of the library so it is rather a runtime dependency ?

This is an area where I mostly need someone to tell me if it's "ok" or not.  Here's what I did:
I created a new python package https://github.com/tlambert03/bioformats_jar that _is_ GPL-2.0 Licensed... it contains little more than the jar, a simple way to launch jpype, and a bunch of `pyi` type stubs that give some autocompletion.  each new attempt to wrap bioformats in python seems to either include the jar (python-bioformats, bfio) or download it at runtime (pims), so it seemed like if there was a versioned jar availabled on pypi, they could all just use that and wrap it how they like.  with that, aicsimageio isn't distributing the jar at all... just requiring the `bioformats_jar` to be installed at runtime.

curious to hear your thoughts on that

> do you have a timeline for this functionality to be released and a general feeling for how experimental/supported the feature will be? The question of Bio-Formats in Python is regularly asked e.g. on image.sc. I assume on the OME side we would be more than happy to point at `aicsimageio` with this change integrated as one of the solutions but given the exposure I would definitely not do that without your consent ;)

i'm hopeful that this will be a robust solution that offers both ease of installation and features like lazy-loading (I've used all the existing solutions and have a decent understanding of their strengths and weaknesses) ... but I'm also no longer naïve enough to think that we won't hit some initial difficulties :joy:
I suspect the main issues will be around platform-dependent specifics of the jpype connection and java installation (I'll generally suggest people use `conda install -c conda-forge openjdk`... but we'll see what else works)

toloudis commented on aicsimageio/metadata/utils.py:
```diff
@@ -564,3 +564,134 @@ def ome_to_numpy_dtype(ome_type: PixelType) -> np.dtype:
+    coords: Dict[str, Union[List[Any], Union[types.ArrayLike, Any]]]
+        The coordinate planes / data for each dimension.
+    """
+    import numpy as np
```
numpy is already imported here

toloudis commented on aicsimageio/readers/bioformats_reader.py:
```diff
@@ -0,0 +1,442 @@
+from ..utils.dask_proxy import DaskArrayProxy
+from .reader import Reader
+
+if TYPE_CHECKING:
```
I'm just curious, why are these only imported if TYPE_CHECKING?

tlambert03 replied:
this again is just a habit (and I'm happy to move them all up to the top if it feels jarring to have this in some files and not in others.  apologies there).  The rationale is twofold:
1) i think putting things in TYPE_CHECKING is a nice, clear way to say "nothing here is needed at runtime... all of these imports will only appear in type annotations."
2) it helps to reduce the possibility of future circular imports.  that doesn't seem to be something that this library is in any danger any time soon, but in napari I've certainly had to go back and hide imports that had previously been imported at the tope just for the sake of type checking... hence the habit.

but like i said: i can see that it's a bit weird to have it sometimes but not always... happy to move it back up, shall i?

toloudis replied:
understood.  I don't have a strong preference, but a mild one for moving it back up. Happy to let Jackson weigh in and don't want to just add meaningless busywork.

JacksonMaxfield replied:
I actually like the TYPE_CHECKING thing personally. It's definitely "safer" imo.

toloudis replied:
Then I'm good with it.

toloudis commented on aicsimageio/readers/bioformats_reader.py:
```diff
@@ -0,0 +1,442 @@
+        try:
+            with LociFile(self._path) as rdr:
+                self._scenes: Tuple[str, ...] = tuple(
+                    metadata_utils.generate_ome_image_id(i)
```
I think the scene names could actually come from the metadata, no?  Now I'm second guessing, do we want to enforce that they are these ome_image_ids or the names coming from the file?

tlambert03 replied:
yeah, wasn't quite sure what to do here either!  happy to take your advice on that one.  @JacksonMaxfield?

toloudis replied:
I had to think for a sec, but as a user I might already know the scene names I put in the microscopy file and so might want to reference them as such.

JacksonMaxfield replied:
If there is an easy loci one liner then yes I would personally pull the scene names out from the raw file.

tlambert03 replied:
> If there is an easy loci one liner then yes I would personally pull the scene names out from the raw file.

so just to confirm, do you guys mean the "name" field of each top level Image (for OME).  or the proprietary scene name?  If we take, for example, the `s_3_t_1_c_3_z_5.czi` test file (where the "CziReader" would yield scenes named "P1, P2, P3") ...  the Image "Name" attributes look like this:

```xml
<Image ID="Image:0" Name="s_3_t_1_c_3_z_5.czi #1">
<Image ID="Image:1" Name="s_3_t_1_c_3_z_5.czi #2">
<Image ID="Image:1" Name="s_3_t_1_c_3_z_5.czi #3">
```
(on leica, those Name fields actually _do_ match what you extract for scene names)

the zeiss P1/2/3 strings are buried in the original metadata structured annotations (which would be a pain to extract for each format):

```xml
        <XMLAnnotation ID="Annotation:668" Namespace="openmicroscopy.org/OriginalMetadata">
            <Value>
                <OriginalMetadata>
                    <Key>Information|Image|S|Scene|Name</Key>
                    <Value>[P2, P3, P1]</Value>
                </OriginalMetadata>
            </Value>
        </XMLAnnotation>
```


so should we A) just use `Image:0` etc... or B) use the `Name` of the OME Image attribute?

toloudis replied:
Option B seems the most likely to be what the user would have named their scenes, I would hope?

toloudis replied:
(Does bioformats always put something fake there or is it at least attempting to fish out the real names?)

JacksonMaxfield replied:
Yea This has bee a bad spot in the API.. I think some readers currently return what is the "ID" and some readers currently return what is the "name" here. Personally, I am fine with returning the ID (as you have done) as I think a majority of our readers currently return ID... We should make a decision on that later though.

If @toloudis agrees with my above sentiment I can file an issue for us to standardize our ID or Name selection later.

toloudis replied:
My sense is that Name is a better guess for what a user would pass as input, and a more sensible scene name for our API to return.   Those ID things are more like an integer index and if bioformats is actually extracting name info from the PFF's metadata, then it would put it in the Name field and not the ID field.

toloudis replied:
I think the only implication on this is that you have to get the ome metadata here, but that is no worse than what other readers would have to do.

JacksonMaxfield replied:
That's fair. I guess using name does make more sense. Even if the names are gross like the example you provided.

tlambert03 replied:
will do

tlambert03 replied:
updated.  they _are_ pretty gross 😂 ... (except for leica, which does get better)

toloudis commented on aicsimageio/readers/bioformats_reader.py:
```diff
@@ -0,0 +1,442 @@
+        threshold for memoizing the reader. If the time required to call
+        `reader.setId()` is larger than this number, the initialized reader
+        (including all reader wrappers) will be cached in a memo file for later
+        retrieval. by default 50.  set to 0 or less to turn off memoization.
```
what are the units of this number?

tlambert03 replied:
milliseconds.  adding a comment.
Actually, I want to generally draw your attention here.  This is a nice feature, but I think it has to be off by default: it writes a memo file to the same directory as the file being opened (or the location of the BIOFORMATS_MEMO_DIR if set).  While it will reduce the time to load the file on future loads, it can actually be a couple megabytes depending on the size of the original file.  So, just wanted you to notice it's there (particularly if we find that we need to go back to a dask mode where the file is opened and closed for every chunk)... but it's off for now

toloudis replied:
I agree it should be off by default. (We could add a reader kwarg to forward to this if it makes sense for users to want to control it) What does it do if you don't have write permission?  Is it an error or does it just silently not memoize?

tlambert03 replied:
> Is it an error or does it just silently not memoize?

it's a good question!  I suspect it errors, but I haven been able to confirm.  something for a future PR

sbesson commented:
Definitely worth double checking with Allen Legal but my understanding is that as long as this library does not ship a copy or a derivative of the Bio-Formats GPL JAR under a more permissive license, it should be fine. This means Talley's wrapper class itself is distributed and licensed under the terms of this library but Bio-Formats is a runtime dependency.

And obviously, :+1: to making open community format readers/writers in a license as permissive as possible.

Happy to follow your lead regarding the advertisement of the functionality. Eventually, it feels like this could be mentioned on the Bio-Formats documentation for people interested in Python - similarly to https://github.com/aoles/RBioFormats for instance.

For the ease of distribution, the conda package is an interesting idea especially if the additional bindings requires platform-specific instructions.

tlambert03 commented:
ok. I just dumped the list of supported formats to the end of `formats.py` (after brief chat with @JacksonMaxfield yesterday).   I commented out everything that has a duplicate entry above.  It's... a lot to digest 😂 .  I think that we need #307 to really address this issue at this point.  but it's a start, and I added a few general AICSImage tests as well

JacksonMaxfield commented on aicsimageio/formats.py:
```diff
@@ -194,4 +194,183 @@
     "lfr": "aicsimageio.readers.default_reader.DefaultReader",
     "lfp": "aicsimageio.readers.default_reader.DefaultReader",
     "spe": "aicsimageio.readers.default_reader.DefaultReader",
+    # Bioformats
+    # note: some of these should probably be above DefaultReader...
+    # but we need https://github.com/AllenCellModeling/aicsimageio/issues/307
```
@toloudis I am fine with merging this and then fixing / updating when I make a PR for #307 on Friday

JacksonMaxfield commented on aicsimageio/readers/bioformats_reader.py:
```diff
@@ -0,0 +1,463 @@
+        except jpype.JVMNotFoundException:
+            raise
+        except Exception as e:
+            print(e)
```
Swap to log.error or remove imo.

JacksonMaxfield commented on aicsimageio/tests/expectations.json:
```diff
@@ -0,0 +1,38 @@
+{
```
Was this meant to be committed? I don't see a reference to this file anywhere...

tlambert03 replied:
whoops... nope sorry. something I was playing with.  perhaps for a later PR :)

JacksonMaxfield replied:
I think I know what you were trying to do and ya..... we definitely need a central file where all these expectations are kept in cases like this were multiple readers or tests are doing largely the same checks. :joy:

JacksonMaxfield replied:
Later PR though.

JacksonMaxfield reviewed (approved):
All good from my end.

I have some minor questions but otherwise I think this is ready to go.

Last nitpick: add this reader to the README

tlambert03 commented:
updated readme, just added this to the "supports reading" section:

> - Files supported by [Bio-Formats](https://docs.openmicroscopy.org/bio-formats/latest/supported-formats.html) -- (`pip install aicsimageio[bioformats]`)

MatteBailey reviewed (approved):
Very cool, good work @tlambert03!

toloudis reviewed (approved):
I feel like we should also get #307 implemented before doing a version bump and release.

JacksonMaxfield commented:
> I feel like we should also get #307 implemented before doing a version bump and release.

Agree. I think there a couple of small PRs I will be making on Friday.

Assume, PRs on Friday, quick reviews, release 4.1 on Monday

tlambert03 commented:
thanks for all the input everyone!  It's nice to be a new contributor here :)

JacksonMaxfield commented:
Thank you! Fast and incredible work on this. Learned a lot myself.
